packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -15,137 +15,275 @@ const { details: X, quote: q } = assert;
+ * @returns {Facetiousness} an assessment of the facetiousness of `obj`
  */
-export function makeCache(size, fetch, store) {
-  let lruHead;
-  let lruTail;
-  let dirtyCount = 0;
-  const liveTable = new Map();
-
-  const cache = {
-    makeRoom() {
-      while (liveTable.size > size && lruTail) {
-        // kdebug(`### vo LRU evict ${lruTail.baseRef} (dirty=${lruTail.dirty})`);
-        liveTable.delete(lruTail.baseRef);
-        if (lruTail.dirty) {
-          store(lruTail.baseRef, lruTail.rawState);
-          lruTail.dirty = false;
-          dirtyCount -= 1;
-        }
-        lruTail.rawState = null;
-        if (lruTail.prev) {
-          lruTail.prev.next = undefined;
-        } else {
-          lruHead = undefined;
-        }
-        const deadEntry = lruTail;
-        lruTail = lruTail.prev;
-        deadEntry.next = undefined;
-        deadEntry.prev = undefined;
-      }
-    },
-    markDirty(entry) {
-      if (!entry.dirty) {
-        entry.dirty = true;
-        dirtyCount += 1;
-      }
-    },
-    setSize(newSize) {
-      if (newSize < size) {
-        size = newSize;
-        cache.makeRoom();
-      } else {
-        size = newSize;
-      }
-    },
-    flush() {
-      if (dirtyCount > 0) {
-        let entry = lruTail;
-        while (entry) {
-          if (entry.dirty) {
-            store(entry.baseRef, entry.rawState);
-            entry.dirty = false;
-          }
-          entry = entry.prev;
-        }
-        dirtyCount = 0;
-      }
-    },
-    remember(innerObj) {
-      if (liveTable.has(innerObj.baseRef)) {
-        return;
-      }
-      cache.makeRoom();
-      liveTable.set(innerObj.baseRef, innerObj);
-      innerObj.prev = undefined;
-      innerObj.next = lruHead;
-      if (lruHead) {
-        lruHead.prev = innerObj;
-      }
-      lruHead = innerObj;
-      if (!lruTail) {
-        lruTail = innerObj;
-      }
-      // kdebug(`### vo LRU remember ${lruHead.baseRef}`);
-    },
-    refresh(innerObj) {
-      if (innerObj !== lruHead) {
-        const oldPrev = innerObj.prev;
-        const oldNext = innerObj.next;
-        if (oldPrev) {
-          oldPrev.next = oldNext;
-        } else {
-          lruHead = oldNext;
-        }
-        if (oldNext) {
-          oldNext.prev = oldPrev;
-        } else {
-          lruTail = oldPrev;
-        }
-        innerObj.prev = undefined;
-        innerObj.next = lruHead;
-        lruHead.prev = innerObj;
-        lruHead = innerObj;
-        // kdebug(`### vo LRU refresh ${lruHead.baseRef}`);
-      }
-    },
-    lookup(baseRef, load) {
-      let innerObj = liveTable.get(baseRef);
-      if (innerObj) {
-        cache.refresh(innerObj);
-      } else {
-        innerObj = { baseRef, rawState: null, repCount: 0 };
-        cache.remember(innerObj);
-      }
-      if (load && !innerObj.rawState) {
-        innerObj.rawState = fetch(baseRef);
+function assessFacetiousness(obj, inner) {
```

Comment:
Sounds good, but I think we can simplify it further by bailing out upon the very first `not`, so:
```js
function assessFacetiousness(obj, inner) {
  if (typeof obj !== 'object') {
    return 'not';
  }
  let result;
  for (const prop of Reflect.ownKeys(obj)) {
    const value = obj[prop];
    // for code brevity, default to a negative presumption
    let resultFromProp = 'not';
    if (typeof value === 'function') {
      resultFromProp = 'one';
    } else if (
      !inner &&
      // symbols are not permitted as facet names
      typeof prop !== 'symbol' &&
      assessFacetiousness(value, true) === 'one'
    ) {
      resultFromProp = 'many';
    } else {
      return 'not';
    }
    // capture the result of inspecting the first property
    // and bail out upon encountering any deviation
    if (!result) {
      result = resultFromProp;
    } else if (resultFromProp !== result) {
      return 'not';
    }
  }
  // empty objects are methodless Far objects
  return /** @type {Facetiousness} */ (result || 'one');
}
```

Reply:
also, it's high time we had a unit test of this particular function, so I'll export it and write a more focussed test

Reply:
oh, and in fact `inner` isn't required, we can allow the recursion and then notice that the result is `multi` instead of `one` and that disqualifies the facet.

I'll break this whole function out to a separate file while I'm at it, since `virtualObjectManager.js` is a bit too large.